### PR TITLE
NAS-131651 / Add integrity streams support

### PR DESCRIPTION
**Pull request by anodos325**
This commit adds basic wrappers around the FSCTLs to get and set integrity streams. These are a ReFS feature that is checked by VEEAM to provide fast copy support. For now the implementation is minimal (just providing basic information about whether checksumming is enabled on the dataset). If user attempts to set it to something other than some variant of enabled, then the attempt will fail with ACCESS_DENIED. So far in testing presenting the ZFS record size as the volume cluster size has not presented issues for clients, but further evaluation may be required. This feature is currently off-by-default.

### CODE REVIEW

File: libcli/smb/smb_constants.h [hunk 1]
```diff
@@ -552,6 +553,7 @@ enum csc_policy {
 #define FSCTL_FILE_LEVEL_TRIM		(FSCTL_FILESYSTEM | FSCTL_ACCESS_WRITE | 0x0208 | FSCTL_METHOD_BUFFERED)
 #define FSCTL_OFFLOAD_READ		(FSCTL_FILESYSTEM | FSCTL_ACCESS_READ | 0x0264 | FSCTL_METHOD_BUFFERED)
 #define FSCTL_OFFLOAD_WRITE		(FSCTL_FILESYSTEM | FSCTL_ACCESS_WRITE | 0x0268 | FSCTL_METHOD_BUFFERED)
+#define FSCTL_GET_INTEGRITY_INFORMATION 0x0009027C
```
Comment (mgrimesix) [inline libcli/smb/smb_constants.h]: Why not cobble this together in the same manner as the others?

Comment (anodos325) [reply libcli/smb/smb_constants.h]: This precise value is documented in MS-FSCC and in kernel SMB client / server code. I think cobbling together is less clear. There are other FSCTLs defined via hex values FWIW.

**Review by mgrimesix — APPROVED** [review]
I don't see any issues. 
Approving.

File: source3/modules/smb_libzfs.c [hunk 2]
```diff
@@ -614,10 +614,13 @@ static struct zfs_dataset *copy_to_external(TALLOC_CTX *mem_ctx,
 		if (out->properties == NULL) {
 			TALLOC_FREE(out);
 			errno = ENOMEM;
+			DS_UNLOCK();
```
Comment (bmeagherix) [inline source3/modules/smb_libzfs.c:617]: Good catch.

File: source3/modules/vfs_zfs_core.c [hunk 3]
```diff
@@ -748,6 +686,9 @@ static int zfs_core_connect(struct vfs_handle_struct *handle,
 
 		config->zfs_quota_enabled = lp_parm_bool(SNUM(handle->conn),
 				"zfs_core", "zfs_quota_enabled", true);
+
+		config->zfs_integrity_streams_enabled = lp_parm_bool(SNUM(handle->conn),
+				"zfs_core", "zfs_integrity_streams", false);
```
Comment (bmeagherix) [inline source3/modules/vfs_zfs_core.c:691]: Do we need to document these `zfs_core` samba parameters somewhere?

File: source3/modules/vfs_zfs_core_fsctl.c [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+	uint64_t record_size;
+};
+
+static bool zfs_core_get_integrity_info(struct vfs_handle_struct *handle,
+					struct files_struct *fsp,
+					struct zfs_integrity_info *info)
+{
+	struct zfs_dataset *ds = NULL;
+	struct zfs_core_config_data *config = NULL;
+
```
Comment (bmeagherix) [inline source3/modules/vfs_zfs_core_fsctl.c:43]: Maybe assert that info is not null.

File: source3/modules/vfs_zfs_core_fsctl.c [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	struct integrity_state integ_state;
+	struct zfs_integrity_info info;
+
+	if (!zfs_core_get_integrity_info(handle, fsp, &info)) {
+		return NT_STATUS_INVALID_DEVICE_REQUEST;
+	}
+
+	integ_state = (struct integrity_state){
+		.reserved = 0,
+		.chksum_chunk_sz = 4096,
```
Comment (bmeagherix) [inline source3/modules/vfs_zfs_core_fsctl.c:160]: Might be nice to comment wrt 4096.

File: librpc/idl/ioctl.idl [hunk 6]
```diff
@@ -240,3 +240,30 @@ interface fsctl
...
+	} fsctl_set_integrity_req;
+
+	/* MS-FSCC 2.3.19 GET_INTEGRITY_INFORMATION */
+	typedef [public] struct {
+		uint16 chksum_algo;
+		uint16 reserved;
+		uint32 flags;
+		uint32 chksum_chunk_sz;
+		uint32 cluster_sz;
+	} integrity_state;
```
Comment (bmeagherix) [inline librpc/idl/ioctl.idl:268]: Why `_state` rather than `_info`? (Meh)

**Review by bmeagherix — APPROVED** [review]

File: librpc/idl/ioctl.idl (hunk 6 above)
Comment (anodos325) [reply librpc/idl/ioctl.idl:268]: _state is consistent with other naming in file. cf. compression_state.

File: source3/modules/vfs_zfs_core.c (hunk 3 above)
Comment (anodos325) [reply source3/modules/vfs_zfs_core.c:691]: I don't expect these to be modified by end-users, but I'll add writing a manpage to the TODO list.

**Pull request merged**